gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageStubProvider.java

```diff
@@ -34,9 +50,82 @@
 
   private final GoogleCloudStorageReadOptions readOptions;
   private final ExecutorService backgroundTasksThreadPool;
-  private final List<ManagedChannel> mediaChannelPool;
+  private final List<ChannelAndRequestCounter> mediaChannelPool;
 
-  private int nextChannel = 0;
+  // An interceptor that can be added around a gRPC channel which keeps a count of the number
+  // of requests that are active at any given moment.
+  final class ActiveRequestCounter implements ClientInterceptor {
+
+    // A count of the number of RPCs currently underway for one gRPC channel channel.
+    private AtomicInteger ongoingRequestCount;
+
+    public ActiveRequestCounter() {
+      ongoingRequestCount = new AtomicInteger(0);
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+        MethodDescriptor<ReqT, RespT> methodDescriptor, CallOptions callOptions, Channel channel) {
+      ClientCall<ReqT, RespT> newCall = channel.newCall(methodDescriptor, callOptions);
+      final AtomicBoolean countedCancel = new AtomicBoolean(false);
+
+      // A streaming call might be terminated in one of several possible ways:
+      // * The call completes normally -> onClose() will be invoked.
+      // * The context is cancelled -> CancellationListener.cancelled() will be called.
+      // * The call itself is cancelled (doesn't currently happen) -> ClientCall.cancel() called.
+      //
+      // It's possible more than one of these could happen, so we use countedCancel to make sure we
+      // don't double count a decrement.
+      Context.current().addListener(new CancellationListener() {
+        @Override
+        public void cancelled(Context context) {
+          if(countedCancel.compareAndSet(false, true)) {
+            ongoingRequestCount.decrementAndGet();
+          }
+        }
+      }, backgroundTasksThreadPool);
+
+      return new SimpleForwardingClientCall(newCall) {
+
+        @Override
+        public void cancel(@Nullable String message, @Nullable Throwable cause) {
+          if(countedCancel.compareAndSet(false, true)) {
+            ongoingRequestCount.decrementAndGet();
+          }
+          super.cancel(message, cause);
+        }
+
+        @Override
+        public void start(Listener responseListener, Metadata headers) {
+          ongoingRequestCount.incrementAndGet();
+          this.delegate().start(
+              new SimpleForwardingClientCallListener(responseListener) {
+                @Override
+                public void onClose(Status status, Metadata trailers) {
+                  if(countedCancel.compareAndSet(false, true)) {
+                    ongoingRequestCount.decrementAndGet();
+                  }
+                  super.onClose(status, trailers);
+                }
+              }, headers);
+        }
+      };
+    }
+  }
+
+  class ChannelAndRequestCounter {
```

Comment:
Wouldn't this class make more sense if it's called ChannelWithRequestCounter, a subclass of ManagedChannel with a counter in initializer?  Seems like it's essentially a channel with a counter around it, and you don't need to write something like "channel.channel" down below.

Reply:
Let's address this issue in separate PR.